Refactor prescribeOverrides

vsomayaji opened the pull request:
## Overview

Cleans up the `prescribeOverrides` implementation.

### Notes

With the original implementation, the `TestPrescribeOverrides` test was failing about 10% of the time like this:

```
--- FAIL: TestPrescribeOverrides (0.00s)
	util_test.go:141: 
			Error Trace:	util_test.go:141
			Error:      	Not equal: 
			            	expected: []v2.Service{v2.Service{ID:"test", Name:"test", Description:"test", Tags:[]string(nil), Requires:[]string(nil), Bindable:false, BindingsRetrievable:false, PlanUpdatable:(*bool)(nil), Plans:[]v2.Plan{v2.Plan{ID:"testplan", Name:"testplan", Description:"testplan", Free:(*bool)(nil), Bindable:(*bool)(nil), Metadata:map[string]interface {}(nil), Schemas:(*v2.Schemas)(0xc420387b40)}}, DashboardClient:(*v2.DashboardClient)(nil), Metadata:map[string]interface {}(nil)}}
			            	actual  : []v2.Service{v2.Service{ID:"test", Name:"test", Description:"test", Tags:[]string(nil), Requires:[]string(nil), Bindable:false, BindingsRetrievable:false, PlanUpdatable:(*bool)(nil), Plans:[]v2.Plan{v2.Plan{ID:"testplan", Name:"testplan", Description:"testplan", Free:(*bool)(nil), Bindable:(*bool)(nil), Metadata:map[string]interface {}(nil), Schemas:(*v2.Schemas)(0xc420387af0)}}, DashboardClient:(*v2.DashboardClient)(nil), Metadata:map[string]interface {}(nil)}}
			            	
			            	Diff:
			            	--- Expected
			            	+++ Actual
			            	@@ -34,16 +34,3 @@
			            	       }),
			            	-      Update: (*v2.InputParametersSchema)({
			            	-       Parameters: (map[string]interface {}) (len=4) {
			            	-        (string) (len=7) "$schema": (string) (len=39) "http://json-schema.org/draft-06/schema#",
			            	-        (string) (len=10) "properties": (map[string]interface {}) (len=1) {
			            	-         (string) (len=9) "req_param": (map[string]interface {}) (len=1) {
			            	-          (string) (len=4) "type": (string) (len=6) "string"
			            	-         }
			            	-        },
			            	-        (string) (len=8) "required": ([]string) (len=1) {
			            	-         (string) (len=9) "req_param"
			            	-        },
			            	-        (string) (len=4) "type": (string) (len=6) "object"
			            	-       }
			            	-      })
			            	+      Update: (*v2.InputParametersSchema)(<nil>)
			            	      }),
			Test:       	TestPrescribeOverrides
			Messages:   	override_param should be removed from Update params too when prescribeOverrides is true
```

Rather than get to the root cause of the test failure, it seemed easier to just address the TODO and refactor `prescribeOverrides`. Now, instead of creating new schemas, we just remove the overridden keys from the existing schemas. 

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jaymccon commented on pkg/broker/util.go:
```diff
@@ -31,78 +31,37 @@ func getGlobalOverrides(brokerID string) map[string]string {
-						"properties": props,
-						"$schema":    "http://json-schema.org/draft-06/schema#",
-					}
-					if len(required) > 0 {
```
cloudfoundry insists on required containing more than 0 slices, so we need to check whether this is empty after removing all the overrides and delete the key if it is.

jaymccon replied:
I'll add a test for this in the next couple of days

vsomayaji replied:
@jaymccon does Cloud Foundry care if `properties` is an empty map? Or do we have to delete that too if all properties have been overridden?

jaymccon replied:
having an empty properties block is accepted as valid by CF.

tested with public S3 production plan, aws_access_key, aws_secret_key, region overridden. 

GET /v2/catalog returns:

```json
...
{
  "service_instance": {
    "create": {
      "parameters": {
        "$schema": "http://json-schema.org/draft-06/schema#",
        "properties": {},
        "type": "object"
      }
    }
  }
}
...
```

vsomayaji replied:
Sweet. I added the bit for removing `required` if all required params have been overridden. I also added a test.

vsomayaji commented:
@jaymccon strangely, it does seem to address the intermittent test failures that we've been experiencing. When running the tests against `master` 100 times, they failed 16 times. When running them against this branch, they didn't fail. 😄 
```
$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
$ failures=0
$ for i in $(seq 1 100); do if ! go test --count 1 ./... > /dev/null; then failures=$((failures+1)); fi; done
$ echo failures=$failures
failures=16

$ git checkout refactor-prescribeOverrides
Switched to branch 'refactor-prescribeOverrides'
Your branch is up-to-date with 'origin/refactor-prescribeOverrides'.
$ failures=0
$ for i in $(seq 1 100); do if ! go test --count 1 ./... > /dev/null; then failures=$((failures+1)); fi; done
$ echo failures=$failures
failures=0
```
